### PR TITLE
/custom edit and video embed

### PR DESCRIPTION
**Pull request by BIRD311**
As discussed with fingerprince "add as custom command" will ignore the video embeds now and I added /custom edit

"add as custom command" will take the embed content if you use it on a bot /say response. So you can basically copy commands (could be usefull if you break something with edit).
The value in "/custom edit" is optional and it will remove the value if left blank.
Ignoring the video embed (or rather, there shouldn't be one in the first place or you get the url in the message twice) means you also have to add the thumbnail.
I'm sure you can break something with bad urls but since this is a moderator command I didn't check too much.

### CODE REVIEW

File: MomentumDiscordBot/Commands/General/GeneralModule.cs [hunk 1]
```diff
@@ -13,19 +15,39 @@ public class GeneralModule : MomentumModuleBase
...
                     Description = command.Description,
                     Color = MomentumColor.Blue
-                }.Build();
-                await context.CreateResponseAsync(embed: embed);
+                };
+                if (Uri.IsWellFormedUriString(command.ThumbnailUrl, UriKind.Absolute))
+                {
+                    embedBuilder.Thumbnail = new DiscordEmbedBuilder.EmbedThumbnail
+                    {
+                        Url = command.ThumbnailUrl.ToString(),
```
Comment (tsa96) [inline MomentumDiscordBot/Commands/General/GeneralModule.cs]: ToString is redundant

File: MomentumDiscordBot/Commands/General/GeneralModule.cs [hunk 2]
```diff
@@ -13,19 +15,39 @@ public class GeneralModule : MomentumModuleBase
...
+                        Height = 90,
+                        Width = 160
+                    };
+                }
+                var message = new DiscordMessageBuilder()
+                                .AddEmbed(embedBuilder.Build());
+
+                if (Uri.IsWellFormedUriString(command.ButtonUrl, UriKind.Absolute))
+                    message.AddComponents(new DiscordLinkButtonComponent(command.ButtonUrl.ToString(), command.ButtonLabel ?? "link"));
```
Comment (tsa96) [inline MomentumDiscordBot/Commands/General/GeneralModule.cs]: Another redundant ToString

File: MomentumDiscordBot/Commands/Autocomplete/CustomCommandAutoCompleteProvider.cs [hunk 3]
```diff
@@ -29,4 +29,14 @@ public Task<IEnumerable<DiscordAutoCompleteChoice>> Provider(AutocompleteContext
             return Task.FromResult(choices.Select(command => new DiscordAutoCompleteChoice(command, command)));
         }
     }
+
+    public class CustomCommandPropertyChoiceProvider : IChoiceProvider
+    {
+        public Task<IEnumerable<DiscordApplicationCommandOptionChoice>> Provider()
+        {
+            var properties = typeof(CustomCommand).GetProperties();
+            var choices = properties.Take(25).Select(poperty => new DiscordApplicationCommandOptionChoice(poperty.Name, poperty.Name));
```
Comment (tsa96) [inline MomentumDiscordBot/Commands/Autocomplete/CustomCommandAutoCompleteProvider.cs]: Typo in `popertry`. Doesn't really matter, but may as well change it if we're fixing up some other stuff.

File: MomentumDiscordBot/Commands/Moderator/ModeratorCustomModule.cs [hunk 4]
```diff
@@ -21,29 +23,43 @@ public async Task MessageMenu(ContextMenuContext context)
...
+            string buttonLabel = null;
+            string thumbnailUrl = null;
+
+            if (message.Interaction is { Name: GeneralModule.SayCommandName })
             {
-                title = "";
-                description = message.Content;
+                if (message.Embeds.Any())
+                {
+                    var embed = message.Embeds.First();
```
Comment (tsa96) [inline MomentumDiscordBot/Commands/Moderator/ModeratorCustomModule.cs]: I don't think it's correct practice to use LINQ here where we can just use an index.
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1826

**Review by tsa96 — CHANGES_REQUESTED** [review]
creationTimestamp throws an unhandled exception if the timestamp isn't valid, that should be caught. Though it doesn't seem like a very useful property anyway (if anything I just get confused over what it does, and I can't see us ever caring about when the command is registered), I think it'll be best to just remove that.

File: MomentumDiscordBot/Commands/Moderator/ModeratorCustomModule.cs (hunk 4 above)
Comment (BIRD311) [reply MomentumDiscordBot/Commands/Moderator/ModeratorCustomModule.cs]: Did you IDE give you a warning here? There are multiple instances where First() is used like this in the code so I should probably change all of them?

Comment (tsa96) [reply MomentumDiscordBot/Commands/Moderator/ModeratorCustomModule.cs]: Yeah, though I was only reading through your new code. If it's happening in other places it's not a big deal, I've been trying to only flag stuff I've noticed in new code, the same stuff is used in other places then we can let it slide, there's lots of small issues that linters spot in the code already that we don't need to worry about.

Comment (BIRD311) [reply MomentumDiscordBot/Commands/Moderator/ModeratorCustomModule.cs]: ok I'll change this and maybe I can clean up the rest of the code later

**Review by tsa96 — APPROVED** [review]
This is all working well, nice job!

**Pull request merged**